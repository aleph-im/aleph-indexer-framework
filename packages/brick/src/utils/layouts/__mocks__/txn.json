{
  "blockTime": 1667836100,
  "meta": {
    "err": null,
    "fee": 10000,
    "innerInstructions": [
      {
        "index": 1,
        "instructions": [
          {
            "parsed": {
              "info": {
                "account": "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
                "amount": "1165171360",
                "authority": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
                "mint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So"
              },
              "type": "burn"
            },
            "program": "spl-token",
            "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
          }
        ]
      }
    ],
    "logMessages": [
      "Program 11111111111111111111111111111111 invoke [1]",
      "Program 11111111111111111111111111111111 success",
      "Program MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD invoke [1]",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [2]",
      "Program log: Instruction: Burn",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 4800 of 382342 compute units",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success",
      "Program MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD consumed 29711 of 400000 compute units",
      "Program MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD success"
    ],
    "postBalances": [
      3844455611, 1503360, 2039280, 19098240, 2094555549, 1, 1141440, 1169280,
      1009200, 934087680
    ],
    "postTokenBalances": [
      {
        "accountIndex": 2,
        "mint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
        "owner": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
        "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
        "uiTokenAmount": {
          "amount": "0",
          "decimals": 9,
          "uiAmount": null,
          "uiAmountString": "0"
        }
      }
    ],
    "preBalances": [
      3845968971, 0, 2039280, 19098240, 2094555549, 1, 1141440, 1169280,
      1009200, 934087680
    ],
    "preTokenBalances": [
      {
        "accountIndex": 2,
        "mint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
        "owner": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
        "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
        "uiTokenAmount": {
          "amount": "1165171360",
          "decimals": 9,
          "uiAmount": 1.16517136,
          "uiAmountString": "1.16517136"
        }
      }
    ],
    "rewards": [],
    "status": {
      "Ok": null
    }
  },
  "slot": 159814698,
  "transaction": {
    "message": {
      "accountKeys": [
        {
          "pubkey": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
          "signer": true,
          "source": "transaction",
          "writable": true
        },
        {
          "pubkey": "DphAzdHtKRxHxXSPjxYPwV5dwhsLVp86r9RnLLcYHLg2",
          "signer": true,
          "source": "transaction",
          "writable": true
        },
        {
          "pubkey": "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
          "signer": false,
          "source": "transaction",
          "writable": true
        },
        {
          "pubkey": "8szGkuLTAux9XMgZ2vtY39jVSowEcpBfFfD8hXSEqdGC",
          "signer": false,
          "source": "transaction",
          "writable": true
        },
        {
          "pubkey": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
          "signer": false,
          "source": "transaction",
          "writable": true
        },
        {
          "pubkey": "11111111111111111111111111111111",
          "signer": false,
          "source": "transaction",
          "writable": false
        },
        {
          "pubkey": "MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD",
          "signer": false,
          "source": "transaction",
          "writable": false
        },
        {
          "pubkey": "SysvarC1ock11111111111111111111111111111111",
          "signer": false,
          "source": "transaction",
          "writable": false
        },
        {
          "pubkey": "SysvarRent111111111111111111111111111111111",
          "signer": false,
          "source": "transaction",
          "writable": false
        },
        {
          "pubkey": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
          "signer": false,
          "source": "transaction",
          "writable": false
        }
      ],
      "addressTableLookups": null,
      "instructions": [
        {
          "parsed": {
            "info": {
              "lamports": 1503360,
              "newAccount": "DphAzdHtKRxHxXSPjxYPwV5dwhsLVp86r9RnLLcYHLg2",
              "owner": "MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD",
              "source": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
              "space": 88
            },
            "type": "createAccount"
          },
          "program": "system",
          "programId": "11111111111111111111111111111111"
        },
        {
          "accounts": [
            "8szGkuLTAux9XMgZ2vtY39jVSowEcpBfFfD8hXSEqdGC",
            "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
            "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
            "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
            "DphAzdHtKRxHxXSPjxYPwV5dwhsLVp86r9RnLLcYHLg2",
            "SysvarC1ock11111111111111111111111111111111",
            "SysvarRent111111111111111111111111111111111",
            "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
          ],
          "data": "D4QoZ9QytqaDY8CFubkmeP",
          "programId": "MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD"
        }
      ],
      "recentBlockhash": "B7L4Ldi21oSy4GYdFxEUzaAhatvgKQhfv4oz3s4HVBrH"
    },
    "signatures": [
      "2XaL2jqiFM69nRGSu3Zxbj955e4cjWT8EduKuXUEsGCXW8zrPiznNS9bpgdNbe1YBGVVRDB9ei9c4swYu2RiGaZ4",
      "3tJWgW5f11Bt1utGvmzBwZgYYP5hzkBvceAyoieR6FL1zAEU3hMie92w1pok1BP9GD23MvfJjAenBaRHmS8kmiA6"
    ]
  }
}
