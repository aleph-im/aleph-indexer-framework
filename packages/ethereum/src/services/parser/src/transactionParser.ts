import { BlockchainId, DefinedParser } from '@aleph-indexer/framework'
import { EthereumClient } from '../../../sdk/client.js'
import { EthereumRawTransaction } from '../../../types.js'
import { EthereumAbiFactory } from './abiFactory.js'
import { EthereumParsedTransaction } from './types.js'

export class EthereumTransactionParser extends DefinedParser<
  EthereumRawTransaction,
  EthereumParsedTransaction
> {
  constructor(
    protected blockchainId: BlockchainId,
    protected abiFactory: EthereumAbiFactory,
    protected ethereumClient: EthereumClient,
  ) {
    super()
  }

  async parse(
    rawTx: EthereumRawTransaction,
  ): Promise<EthereumRawTransaction | EthereumParsedTransaction> {
    if (!rawTx.to) return rawTx

    console.log(`${this.blockchainId} rawTx`, JSON.stringify(rawTx, null, 2))

    const abi = await this.abiFactory.getAbi(rawTx.to)
    const parsedTx = this.ethereumClient.parseTransaction(rawTx, abi)

    console.log(
      `${this.blockchainId} parsed tx => `,
      JSON.stringify(parsedTx, null, 2),
    )
    return parsedTx
  }
}
