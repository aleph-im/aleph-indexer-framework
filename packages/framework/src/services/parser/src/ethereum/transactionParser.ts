import ethers from 'ethers'
import {
  EthereumRawTransaction,
  EthereumParsedTransaction,
} from '@aleph-indexer/core'
import { DefinedParser } from '../base/types.js'
import { Abi, AbiFactory } from './abi/abiFactory.js'

export class EthereumTransactionParser extends DefinedParser<
  EthereumRawTransaction,
  EthereumParsedTransaction
> {
  async parse(
    rawTx: EthereumRawTransaction,
  ): Promise<EthereumRawTransaction | EthereumParsedTransaction> {
    if (!rawTx.to) return rawTx

    const abi: Abi = await AbiFactory.getAbi(rawTx.to)
    const iface = new ethers.utils.Interface(abi as any)

    const parsed = iface.parseTransaction({
      data: rawTx.input,
      value: rawTx.value,
    })

    const parsedTx = { ...rawTx, parsed }

    console.log('ETH parsed => ', parsedTx)
    return parsedTx
  }
}
