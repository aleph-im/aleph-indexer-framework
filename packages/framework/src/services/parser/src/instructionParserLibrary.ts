import { ParsedInstructionV1, RawInstruction } from '@aleph-indexer/core'
import { LayoutFactory } from './layout/layoutFactory.js'
import { DefinedParser } from './parser.js'
import { InstructionParser } from './instructionParser.js'
import { LayoutImplementation } from './layout/types.js'

/**
 * Finds all available instruction parsers and aggregates them for use.
 * It is a strict but optional parser, thus it may return raw instruction if no
 * compatible parser is found.
 */
export class InstructionParserLibrary extends DefinedParser<
  RawInstruction,
  ParsedInstructionV1
> {
  constructor(protected layoutPath?: string) {
    super()
  }

  protected instructionParsers: Record<
    string,
    DefinedParser<RawInstruction, RawInstruction | ParsedInstructionV1>
  > = {}

  /**
   * Parses a raw instruction, if a parser for given solana program is available.
   * @param payload The raw instruction to parse.
   */
  async parse(
    payload: RawInstruction | ParsedInstructionV1,
  ): Promise<RawInstruction | ParsedInstructionV1> {
    const { programId } = payload

    const parser = await this.getParser(programId)
    if (!parser) return payload

    const parsedData = await parser.parse(payload)
    return parsedData as ParsedInstructionV1
  }

  /**
   * Returns the parser for the given programId, if available.
   * @param programId The solana program's address.
   * @protected
   */
  protected async getParser(
    programId: string,
  ): Promise<
    | DefinedParser<RawInstruction, RawInstruction | ParsedInstructionV1>
    | undefined
  > {
    let parser = this.instructionParsers[programId]
    if (parser) return parser

    let implementation: LayoutImplementation
    implementation = await LayoutFactory.getSingleton(programId)

    if (!implementation) {
      const customLayoutsMap = await this.getCustomLayoutsMap()
      implementation = customLayoutsMap[programId]
    }

    if (!implementation) return

    parser = new InstructionParser(
      implementation.programID,
      implementation.name,
      implementation.getInstructionType,
      implementation.accountLayoutMap,
      implementation.dataLayoutMap,
    )

    this.instructionParsers[programId] = parser

    return parser
  }

  protected async getCustomLayoutsMap(): Promise<
    Record<string, LayoutImplementation>
  > {
    if (!this.layoutPath) return {}
    const customLayoutsMap = (await import(this.layoutPath)).default
    return customLayoutsMap || {}
  }
}
