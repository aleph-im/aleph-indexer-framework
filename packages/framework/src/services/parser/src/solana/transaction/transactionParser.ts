import {
  AlephParsedErrorIx,
  AlephParsedInnerTransaction,
  ProgramErrorType,
  SolanaParsedInstruction,
  SolanaParsedTransaction,
  SolanaRawInstruction,
  SolanaRawTransaction,
  TransactionErrorType,
} from '../../../../../types/solana.js'
import { StrictParser } from '../../base/types.js'
import { SolanaInstructionParser } from '../instruction/instructionParser.js'

/**
 * Core class which parses raw transactions into parsed transactions with parsed
 * instructions.
 */
export class SolanaTransactionParser extends StrictParser<
  SolanaRawTransaction,
  SolanaParsedTransaction
> {
  /**
   * @param instructionParserAggregator Aggregates all available instruction parsers for use.
   */
  constructor(protected instructionParserAggregator: SolanaInstructionParser) {
    super()
  }

  /**
   * Parses a raw transaction into a parsed transaction.
   * @param rawTx The raw or partly-parsed transaction to parse.
   */
  async parse(
    rawTx: SolanaRawTransaction | SolanaParsedTransaction,
  ): Promise<SolanaParsedTransaction> {
    if ('parsed' in rawTx) return rawTx as SolanaParsedTransaction
    if (!('transaction' in rawTx))
      throw new Error(`‼️ Empty rawTx ${JSON.stringify(rawTx, null, 2)}`)

    const parsedTx = rawTx as unknown as SolanaParsedTransaction
    parsedTx.parsed =
      rawTx.transaction as unknown as AlephParsedInnerTransaction

    parsedTx.signature = parsedTx.parsed.signatures[0]

    this.parseError(parsedTx)

    if (parsedTx.parsed.message.instructions.length > 0) {
      await this.parseInstructions(parsedTx)
    }

    delete (parsedTx as any).transaction

    return parsedTx
  }

  // @note: Take a look at:
  // https://github.com/solana-labs/solana/blob/master/sdk/src/transaction/error.rs#L13
  // https://github.com/solana-labs/solana/blob/master/sdk/program/src/program_error.rs#L12
  /**
   * Parses the errors of a parsed transaction, if any.
   * @param parsedTx The parsed transaction to parse the errors of.
   * @protected
   */
  protected parseError(parsedTx: SolanaParsedTransaction): void {
    const err = parsedTx.meta?.err as Record<
      string,
      [number, Record<string, number>]
    >

    if (!err) return

    const typeTxt = Object.keys(err)[0]
    const type = TransactionErrorType[
      typeTxt as any
    ] as unknown as TransactionErrorType

    parsedTx.parsed.error = { type }

    if (type === TransactionErrorType.InstructionError) {
      if (!Array.isArray(err.InstructionError)) return

      const [index, subErr] = err[typeTxt]

      const subTypeTxt =
        typeof subErr === 'string' ? subErr : Object.keys(subErr)[0]

      const subType = ProgramErrorType[
        subTypeTxt as any
      ] as unknown as ProgramErrorType

      const ix: AlephParsedErrorIx = {
        index,
        type: subType,
        customCode: subErr[subTypeTxt],
      }

      const logs = parsedTx?.meta?.logMessages || []
      let innerIndex = -2

      for (const log of logs) {
        if (log.indexOf(' failed') !== -1) break
        if (log.indexOf(' invoke') !== -1) innerIndex++
        else if (log.indexOf(' success') !== -1) innerIndex--
      }

      if (innerIndex >= 0) {
        ix.innerIndex = innerIndex
      }

      parsedTx.parsed.error.ix = ix
    }
  }

  /**
   * Parses the instructions of a parsed transaction.
   * @param parsedTx The parsed transaction to parse the instructions of.
   * @protected
   */
  protected async parseInstructions(
    parsedTx: SolanaParsedTransaction,
  ): Promise<void> {
    const instructions = parsedTx.parsed.message.instructions
    const innerInstructions = parsedTx.meta?.innerInstructions || []
    const innerInstructionsMap = innerInstructions.reduce((acc, curr) => {
      acc[curr.index] = curr.instructions
      return acc
    }, {} as Record<number, SolanaRawInstruction[]>)

    const parsedIxs = await Promise.all(
      instructions.map(async (rawIx, index) => {
        const resultIx = (await this.instructionParserAggregator.parse(
          rawIx,
        )) as SolanaParsedInstruction
        resultIx.index = index

        const innerIxs = innerInstructionsMap[index] || []

        if (innerIxs.length > 0) {
          resultIx.innerInstructions = []

          for (const rawIix of innerIxs) {
            const resultIix = (await this.instructionParserAggregator.parse(
              rawIix,
            )) as SolanaParsedInstruction
            resultIix.index = resultIx.innerInstructions.push(resultIix) - 1
          }
        }

        return resultIx
      }),
    )

    parsedTx.parsed.message.instructions = parsedIxs
  }
}
