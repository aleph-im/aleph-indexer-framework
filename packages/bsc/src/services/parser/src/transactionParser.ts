import { EthereumTransactionParser } from '@aleph-indexer/ethereum'
import { BscClient } from '../../../sdk/client.js'
import { BscAbiFactory } from './abiFactory.js'

export class BscTransactionParser extends EthereumTransactionParser {
  constructor(
    protected abiFactory: BscAbiFactory,
    protected client: BscClient,
  ) {
    super(abiFactory, client)
  }
}
