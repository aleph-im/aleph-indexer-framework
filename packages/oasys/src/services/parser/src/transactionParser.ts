import { EthereumTransactionParser } from '@aleph-indexer/ethereum'
import { OasysClient } from '../../../sdk/client.js'
import { OasysAbiFactory } from './abiFactory.js'

export class OasysTransactionParser extends EthereumTransactionParser {
  constructor(
    protected abiFactory: OasysAbiFactory,
    protected client: OasysClient,
  ) {
    super(abiFactory, client)
  }
}
