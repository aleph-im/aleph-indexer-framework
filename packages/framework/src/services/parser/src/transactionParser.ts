import {
  AlephParsedErrorIx,
  ParsedInstructionV1,
  ParsedTransactionV1,
  RawInstruction,
  RawTransactionV1,
  ProgramErrorType,
  TransactionErrorType,
  AlephParsedInnerTransaction,
} from '@aleph-indexer/core'
import { StrictParser } from './parser.js'
import { InstructionParserLibrary } from './instructionParserLibrary.js'

/**
 * Core class which parses raw transactions into parsed transactions with parsed
 * instructions.
 */
export class TransactionParser extends StrictParser<
  RawTransactionV1,
  ParsedTransactionV1
> {
  /**
   * @param instructionParserAggregator Aggregates all available instruction parsers for use.
   */
  constructor(
    protected instructionParserAggregator: InstructionParserLibrary,
  ) {
    super()
  }

  /**
   * Parses a raw transaction into a parsed transaction.
   * @param rawTx The raw or partly-parsed transaction to parse.
   */
  async parse(
    rawTx: RawTransactionV1 | ParsedTransactionV1,
  ): Promise<ParsedTransactionV1> {
    if ('parsed' in rawTx) return rawTx as ParsedTransactionV1
    if (!('transaction' in rawTx))
      throw new Error(`‼️ Empty rawTx ${JSON.stringify(rawTx, null, 2)}`)

    const parsedTx = rawTx as unknown as ParsedTransactionV1
    parsedTx.parsed =
      rawTx.transaction as unknown as AlephParsedInnerTransaction

    parsedTx.signature = parsedTx.parsed.signatures[0]

    this.parseError(parsedTx)

    if (parsedTx.parsed.message.instructions.length > 0) {
      await this.parseInstructions(parsedTx)
    }

    delete (parsedTx as any).transaction

    return parsedTx
  }

  // @note: Take a look at:
  // https://github.com/solana-labs/solana/blob/master/sdk/src/transaction/error.rs#L13
  // https://github.com/solana-labs/solana/blob/master/sdk/program/src/program_error.rs#L12
  /**
   * Parses the errors of a parsed transaction, if any.
   * @param parsedTx The parsed transaction to parse the errors of.
   * @protected
   */
  protected parseError(parsedTx: ParsedTransactionV1): void {
    const err = parsedTx.meta?.err as Record<
      string,
      [number, Record<string, number>]
    >

    if (!err) return

    const typeTxt = Object.keys(err)[0]
    const type = TransactionErrorType[
      typeTxt as any
    ] as unknown as TransactionErrorType

    parsedTx.parsed.error = { type }

    if (type === TransactionErrorType.InstructionError) {
      if (!Array.isArray(err.InstructionError)) return

      const [index, subErr] = err[typeTxt]

      const subTypeTxt =
        typeof subErr === 'string' ? subErr : Object.keys(subErr)[0]

      const subType = ProgramErrorType[
        subTypeTxt as any
      ] as unknown as ProgramErrorType

      const ix: AlephParsedErrorIx = {
        index,
        type: subType,
        customCode: subErr[subTypeTxt],
      }

      const logs = parsedTx?.meta?.logMessages || []
      let innerIndex = -2

      for (const log of logs) {
        if (log.indexOf(' failed') !== -1) break
        if (log.indexOf(' invoke') !== -1) innerIndex++
        else if (log.indexOf(' success') !== -1) innerIndex--
      }

      if (innerIndex >= 0) {
        ix.innerIndex = innerIndex
      }

      parsedTx.parsed.error.ix = ix
    }
  }

  /**
   * Parses the instructions of a parsed transaction.
   * @param parsedTx The parsed transaction to parse the instructions of.
   * @protected
   */
  protected async parseInstructions(parsedTx: ParsedTransactionV1): Promise<void> {
    const instructions = parsedTx.parsed.message.instructions
    const innerInstructions = parsedTx.meta?.innerInstructions || []
    const innerInstructionsMap = innerInstructions.reduce((acc, curr) => {
      acc[curr.index] = curr.instructions
      return acc
    }, {} as Record<number, RawInstruction[]>)

    const parsedIxs = await Promise.all(
      instructions.map(async (rawIx, index) => {
        const resultIx = (await this.instructionParserAggregator.parse(
          rawIx,
        )) as ParsedInstructionV1
        resultIx.index = index

        const innerIxs = innerInstructionsMap[index] || []

        if (innerIxs.length > 0) {
          resultIx.innerInstructions = []

          for (const rawIix of innerIxs) {
            const resultIix = (await this.instructionParserAggregator.parse(
              rawIix,
            )) as ParsedInstructionV1
            resultIix.index = resultIx.innerInstructions.push(resultIix) - 1
          }
        }

        return resultIx
      }),
    )

    parsedTx.parsed.message.instructions = parsedIxs
  }
}
