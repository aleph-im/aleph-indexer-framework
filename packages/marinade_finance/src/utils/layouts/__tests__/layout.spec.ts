import {InstructionParserLibrary} from "@aleph-indexer/framework/dist/src/services/parser/src/instructionParserLibrary.js";
import path from "path";
import {TransactionParser} from "@aleph-indexer/framework/dist/src/services/parser/src/transactionParser.js";
import txn from "../__mocks__/txn.json";
import {fileURLToPath} from "url";
import {AlephParsedParsedInstruction} from "@aleph-indexer/core";

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

describe('TransactionParser', () => {
  it('parses a transaction', async () => {
    const layoutPath = path.join(__dirname, '../layout.js')
    const instructionParserLibrary = new InstructionParserLibrary(layoutPath)
    const transactionParser = new TransactionParser(instructionParserLibrary)
    // @todo: not type safe
    const parsedTxn = await transactionParser.parse(txn as unknown as any)
    console.log(JSON.stringify(parsedTxn, null, 2))
    expect(parsedTxn).toMatchSnapshot()
    const parsedIxn = (parsedTxn.parsed.message.instructions[1] as AlephParsedParsedInstruction).parsed
    expect(parsedIxn.type).toEqual('OrderUnstake')
    console.log(parsedIxn)
  })
})