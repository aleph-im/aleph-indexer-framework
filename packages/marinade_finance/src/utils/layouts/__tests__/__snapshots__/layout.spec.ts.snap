// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionParser parses a transaction 1`] = `
Object {
  "blockTime": 1667836100,
  "meta": Object {
    "err": null,
    "fee": 10000,
    "innerInstructions": Array [
      Object {
        "index": 1,
        "instructions": Array [
          Object {
            "index": 0,
            "parsed": Object {
              "info": Object {
                "account": "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
                "amount": "1165171360",
                "authority": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
                "mint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
              },
              "type": "burn",
            },
            "program": "spl-token",
            "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
          },
        ],
      },
    ],
    "logMessages": Array [
      "Program 11111111111111111111111111111111 invoke [1]",
      "Program 11111111111111111111111111111111 success",
      "Program MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD invoke [1]",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [2]",
      "Program log: Instruction: Burn",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 4800 of 382342 compute units",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success",
      "Program MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD consumed 29711 of 400000 compute units",
      "Program MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD success",
    ],
    "postBalances": Array [
      3844455611,
      1503360,
      2039280,
      19098240,
      2094555549,
      1,
      1141440,
      1169280,
      1009200,
      934087680,
    ],
    "postTokenBalances": Array [
      Object {
        "accountIndex": 2,
        "mint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
        "owner": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
        "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
        "uiTokenAmount": Object {
          "amount": "0",
          "decimals": 9,
          "uiAmount": null,
          "uiAmountString": "0",
        },
      },
    ],
    "preBalances": Array [
      3845968971,
      0,
      2039280,
      19098240,
      2094555549,
      1,
      1141440,
      1169280,
      1009200,
      934087680,
    ],
    "preTokenBalances": Array [
      Object {
        "accountIndex": 2,
        "mint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
        "owner": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
        "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
        "uiTokenAmount": Object {
          "amount": "1165171360",
          "decimals": 9,
          "uiAmount": 1.16517136,
          "uiAmountString": "1.16517136",
        },
      },
    ],
    "rewards": Array [],
    "status": Object {
      "Ok": null,
    },
  },
  "parsed": Object {
    "message": Object {
      "accountKeys": Array [
        Object {
          "pubkey": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
          "signer": true,
          "source": "transaction",
          "writable": true,
        },
        Object {
          "pubkey": "DphAzdHtKRxHxXSPjxYPwV5dwhsLVp86r9RnLLcYHLg2",
          "signer": true,
          "source": "transaction",
          "writable": true,
        },
        Object {
          "pubkey": "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
          "signer": false,
          "source": "transaction",
          "writable": true,
        },
        Object {
          "pubkey": "8szGkuLTAux9XMgZ2vtY39jVSowEcpBfFfD8hXSEqdGC",
          "signer": false,
          "source": "transaction",
          "writable": true,
        },
        Object {
          "pubkey": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
          "signer": false,
          "source": "transaction",
          "writable": true,
        },
        Object {
          "pubkey": "11111111111111111111111111111111",
          "signer": false,
          "source": "transaction",
          "writable": false,
        },
        Object {
          "pubkey": "MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD",
          "signer": false,
          "source": "transaction",
          "writable": false,
        },
        Object {
          "pubkey": "SysvarC1ock11111111111111111111111111111111",
          "signer": false,
          "source": "transaction",
          "writable": false,
        },
        Object {
          "pubkey": "SysvarRent111111111111111111111111111111111",
          "signer": false,
          "source": "transaction",
          "writable": false,
        },
        Object {
          "pubkey": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
          "signer": false,
          "source": "transaction",
          "writable": false,
        },
      ],
      "addressTableLookups": null,
      "instructions": Array [
        Object {
          "index": 0,
          "parsed": Object {
            "info": Object {
              "lamports": 1503360,
              "newAccount": "DphAzdHtKRxHxXSPjxYPwV5dwhsLVp86r9RnLLcYHLg2",
              "owner": "MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD",
              "source": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
              "space": 88,
            },
            "type": "createAccount",
          },
          "program": "system",
          "programId": "11111111111111111111111111111111",
        },
        Object {
          "accounts": Array [
            "8szGkuLTAux9XMgZ2vtY39jVSowEcpBfFfD8hXSEqdGC",
            "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
            "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
            "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
            "DphAzdHtKRxHxXSPjxYPwV5dwhsLVp86r9RnLLcYHLg2",
            "SysvarC1ock11111111111111111111111111111111",
            "SysvarRent111111111111111111111111111111111",
            "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
          ],
          "data": "D4QoZ9QytqaDY8CFubkmeP",
          "index": 1,
          "innerInstructions": Array [
            Object {
              "index": 0,
              "parsed": Object {
                "info": Object {
                  "account": "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
                  "amount": "1165171360",
                  "authority": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
                  "mint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
                },
                "type": "burn",
              },
              "program": "spl-token",
              "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
            },
          ],
          "parsed": Object {
            "info": Object {
              "0": Object {
                "instructionDiscriminator": Array [
                  97,
                  167,
                  144,
                  107,
                  117,
                  190,
                  128,
                  36,
                ],
                "msolAmount": "45731aa0",
              },
              "1": 16,
              "burnMsolAuthority": "BbAHp6ruxUkT9c1v98PuFuuaJQ2YFgcHe1ghGPED4Y3i",
              "burnMsolFrom": "751t72jfqz6fFQuTC9UHzCsx5Lpo1wdEqmHV9dWbKbTW",
              "clock": "SysvarC1ock11111111111111111111111111111111",
              "msolMint": "mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So",
              "newTicketAccount": "DphAzdHtKRxHxXSPjxYPwV5dwhsLVp86r9RnLLcYHLg2",
              "rent": "SysvarRent111111111111111111111111111111111",
              "state": "8szGkuLTAux9XMgZ2vtY39jVSowEcpBfFfD8hXSEqdGC",
              "tokenProgram": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",
            },
            "type": "OrderUnstake",
          },
          "program": "marinade_finance",
          "programId": "MarBmsSgKXdrN1egZf5sqe1TMai9K1rChYNDJgjq7aD",
        },
      ],
      "recentBlockhash": "B7L4Ldi21oSy4GYdFxEUzaAhatvgKQhfv4oz3s4HVBrH",
    },
    "signatures": Array [
      "2XaL2jqiFM69nRGSu3Zxbj955e4cjWT8EduKuXUEsGCXW8zrPiznNS9bpgdNbe1YBGVVRDB9ei9c4swYu2RiGaZ4",
      "3tJWgW5f11Bt1utGvmzBwZgYYP5hzkBvceAyoieR6FL1zAEU3hMie92w1pok1BP9GD23MvfJjAenBaRHmS8kmiA6",
    ],
  },
  "signature": "2XaL2jqiFM69nRGSu3Zxbj955e4cjWT8EduKuXUEsGCXW8zrPiznNS9bpgdNbe1YBGVVRDB9ei9c4swYu2RiGaZ4",
  "slot": 159814698,
}
`;
